Lib/pdb.py

```diff
@@ -1335,29 +1351,85 @@ def do_list(self, arg):
         breaklist = self.get_file_breaks(filename)
         try:
             lines = linecache.getlines(filename, self.curframe.f_globals)
+            instructions = dis.get_instructions(self.curframe.f_code) \
+                            if show_instructions else None
             self._print_lines(lines[first-1:last], first, breaklist,
-                              self.curframe)
+                              self.curframe,
+                              instructions)
             self.lineno = min(last, len(lines))
             if len(lines) < last:
                 self.message('[EOF]')
         except KeyboardInterrupt:
             pass
+
+    def do_list(self, arg):
+        """l(ist) [first [,last] | .]
+
+        List source code for the current file.  Without arguments,
+        list 11 lines around the current line or continue the previous
+        listing.  With . as argument, list 11 lines around the current
+        line.  With one argument, list 11 lines starting at that line.
+        With two arguments, list the given range; if the second
+        argument is less than the first, it is a count.
+
+        The current line in the current frame is indicated by "->".
+        If an exception is being debugged, the line where the
+        exception was originally raised or propagated is indicated by
+        ">>", if it differs from the current line.
+        """
+        self._do_list(arg, False)
     do_l = do_list
 
-    def do_longlist(self, arg):
-        """longlist | ll
-        List the whole source code for the current function or frame.
+    def do_listinst(self, arg):
+        """listinst | li [first[, last] | .]
+
+        List source code for the current file with instructions.
+
+        Without arguments, list 11 lines around the current line or
+        continue the previous listing.  With . as argument, list 11
+        lines around the current line.  With one argument, list 11
+        lines starting at that line. With two arguments, list the
+        given range; if the second argument is less than the first,
+        it is a count.
+
+        The current line in the current frame is indicated by "->".
+        The current instruction is indicated by "-->"
+        If an exception is being debugged, the line where the
+        exception was originally raised or propagated is indicated by
+        ">>", if it differs from the current line.
         """
+        self._do_list(arg, True)
+    do_li = do_listinst
+
+    def _do_longlist(self, arg, show_instructions=False):
         filename = self.curframe.f_code.co_filename
         breaklist = self.get_file_breaks(filename)
         try:
             lines, lineno = inspect.getsourcelines(self.curframe)
         except OSError as err:
             self.error(err)
             return
-        self._print_lines(lines, lineno, breaklist, self.curframe)
+        instructions = dis.get_instructions(self.curframe.f_code) \
```

Comment:
```suggestion
        instructions = dis.get_instructions(self.curframe.f_code, adaptive=True) \
```
See the comment on `_do_list`